### PR TITLE
#3 Reduce Image Size

### LINKED ISSUE #3: Optimize Container Image
Currently, the container image is unnecessarily large because it includes all the build tools for compiling MPI File Utils (MFU) and dependencies. This snippet contains the tools that could be isolated to a build stage https://github.com/NearNodeFlash/nnf-mfu/blob/4f0571fb4ebc3fbc65b122fa6b762b9ef651ad22/Dockerfile#L22-L26

One should use a builder stage in the Dockerfile to compile the necessary libraries and binaries. This stage can be based on the [openmpi-builder](https://github.com/kubeflow/mpi-operator/blob/master/build/base/openmpi-builder.Dockerfile)

After building MFU, another stage can be used as the runner. This stage can be based on [openmpi](https://github.com/kubeflow/mpi-operator/blob/master/build/base/openmpi.Dockerfile) and copy the necessary libraries and binaries to the final image.

The MPI Operator v2 provides a complete example [here](https://github.com/kubeflow/mpi-operator/blob/master/examples/v2beta1/pi/Dockerfile)

### PR DESCRIPTION
**Pull request by NateThornton**
Re-work the Dockerfile to leverage mpioperator's base images. This uses a 'builder' stage to compile MFU and support libs

resolves #3 

Signed-off-by: Nate Roiger <nate.roiger@hpe.com>

### CODE REVIEW

**Comment by NateThornton** [comment]
Base mpioperator image is 185 MB. The latest PR has the image at 208 MB - I don't think we'll get much better than that

File: Dockerfile [hunk 1]
```diff
@@ -33,63 +31,49 @@ RUN apt install -y --no-install-recommends openmpi-bin openssh-server openssh-cl
 RUN mkdir -p /deps
 WORKDIR /deps
 
+RUN mkdir -p /mfu
```
Comment (bdevcich) [inline Dockerfile]: You do this again on line 59. I don't think you need it here.

File: Dockerfile [hunk 2]
```diff
@@ -15,13 +15,11 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM debian:buster
+FROM mpioperator/openmpi-builder AS builder
```
Comment (bdevcich) [inline Dockerfile]: Do we need to tie this to a specific version of the image?

File: Dockerfile [hunk 3]
```diff
@@ -33,63 +31,49 @@ RUN apt install -y --no-install-recommends openmpi-bin openssh-server openssh-cl
...
-
-RUN wget https://github.com/libarchive/libarchive/releases/download/v3.5.3/libarchive-3.5.3.tar.gz \
-    && tar xfz libarchive-3.5.3.tar.gz \
-    && cd libarchive-3.5.3 \
-    && ./configure --prefix=/usr/lib \
+    && ./configure --prefix=/deps/dtcmp/lib --with-lwgrp=/deps/lwgrp/lib \
     && make install
 
+RUN mkdir -p /mfu
 ARG MPI_FILE_UTILS_VERSION="0.11"
```
Comment (bdevcich) [inline Dockerfile:58]: Not sure how much this version changes, but I'd like to see this after the FROM line especially if we use a specific version for the image. Mostly just for clarity.

File: Dockerfile [hunk 4]
```diff
@@ -33,63 +31,49 @@ RUN apt install -y --no-install-recommends openmpi-bin openssh-server openssh-cl
...
-# The following several lines are from the mpi-operator base Dockerfile
-# https://github.com/kubeflow/mpi-operator/blob/master/build/base/Dockerfile
-
-ARG port=2222
-
-# Add priviledge separation directoy to run sshd as root.
-RUN mkdir -p /var/run/sshd
-# Add capability to run sshd as non-root.
-RUN setcap CAP_NET_BIND_SERVICE=+eip /usr/sbin/sshd
+FROM mpioperator/openmpi
```
Comment (bdevcich) [inline Dockerfile]: Same thing here re: version.

File: Dockerfile (hunk 3 above)
Comment (NateThornton) [reply Dockerfile:58]: This doesn't relate to the image. It's a variable for the next line. For locality, I'd like to remain where it currently is.

File: Dockerfile (hunk 2 above)
Comment (NateThornton) [reply Dockerfile]: I'd prefer `latest` at this point in the program. We can lock it down closer to release

File: Dockerfile [hunk 5]
```diff
@@ -30,66 +28,50 @@ RUN apt install -y --no-install-recommends openmpi-bin openssh-server openssh-cl
...
```
Comment (roehrich-hpe) [inline Dockerfile]: Yes, from here on down is awesome, and the only part that really matters.
Nice.

**Review by roehrich-hpe — APPROVED** [review]

File: Dockerfile (hunk 2 above)
Comment (NateThornton) [reply Dockerfile]: Inverted my logic: Fixed the mpioperator images to the 0.3.0 tag; if something comes into latest that we need we can jump to it then (or, more likely, the next published tag)

**Review by bdevcich — APPROVED** [review]

**Pull request merged**